### PR TITLE
SAA-603 implement outbound event for scheduled instance amended

### PR DESCRIPTION
**Pull request by timharrison-moj**
* Entity listener on amend of scheduled instance
* New event for `activities.scheduled-instance.amended`
* This is intended to catch the scenario where a scheduled instance is cancelled or uncancelled.

### CODE REVIEW

File: helm_deploy/hmpps-activities-management-api/values.yaml [hunk 1]
```diff
@@ -35,8 +35,9 @@ generic-service:
     DB_SSL_MODE: "verify-full"
     SCHEDULE_AHEAD_DAYS: 60
     SCHEDULE_AHEAD_DAYS_ONLINE: 14
-    FEATURE_EVENTS_SNS_ENABLED: true
+    FEATURE_EVENTS_SNS_ENABLED: false
     FEATURE_EVENT_ACTIVITIES_ACTIVITY_SCHEDULE_CREATED: true
```
Comment (timharrison-moj) [inline helm_deploy/hmpps-activities-management-api/values.yaml:39]: Correcting a previous issue - FEATURE_EVENTS_SNS_ENABLED should be false in values.yaml, and is overridden in environment-specific files.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import jakarta.persistence.PostUpdate
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.OutboundEvent
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.OutboundEventsService
+
+@Component
+class ScheduledInstanceEntityListener {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt:12]: Just a thought

I do wonder if we could get a way with one listener which then just checks the type and decides what to do e.g.

```
@Component
class EventPublisherListener {

  @Autowired
  private lateinit var outboundEventsService: OutboundEventsService

  companion object {
    private val log: Logger = LoggerFactory.getLogger(this::class.java)
  }

  @PostPersist
  fun onCreate(schedule: Any) {
    when(entity) {
      is Activity -> blah bah
      is ScheduledInstance -> blah blah
    }
  }

  @PostUpdate
  fun onUpdate(entity: Any) {
    when(entity) {
      is Activity -> blah bah
      is ScheduledInstance -> blah blah
    }
  }
  ....
}
```

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt (hunk 2 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt:12]: Interesting - would produce far fewer modules / listeners and send() functions. One for a trial refactor later?

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstanceEntityListener.kt:12]: This is already being done for auditing.

**Pull request merged**